indexing/generate_index.py

```diff
@@ -0,0 +1,82 @@
+# TODO: Remove "contents"
+#       Add links within the doc (contents in a hierarchy?)
+#       Consider auto-gen on release
+
+from collections import defaultdict
+from jinja2 import Environment, FileSystemLoader
+import os
+from pathlib import Path
+from typing import List, NamedTuple
+import warnings
+
+import iris
+
+TEST_DATA_ROOT = Path(__file__).parent / Path("../test_data")
+TEMPLATE_FILE = Path(__file__).parent / Path("INDEX_template.md")
+OUTPUT_FILE = Path(__file__).parent / Path("../INDEX.md")
+
+class FileInfo(NamedTuple):
+    path: Path
+    cube_strs: List[str]
+    warnings: List[str]
+    exceptions: List[str]
+
+def getFilepaths(path : Path) -> List[Path]:
+    path_list = []
+    for root, dirs, files in os.walk(path):
+        for file in files:
+            path_list.append(Path(root) / Path(file))
+    return path_list
+
+def getFileInfos(filepaths: List[Path]) -> List[FileInfo]:
+    file_infos = defaultdict(list)
+
+    for filepath in filepaths:
+        
+        if filepath.suffix in [".py", ".rst", ".txt", ".md"]:
```

Comment:
I used this because it seemed easier to track the file types we don't want to include rather than those we do (otherwise someone adding a file with a novel suffix would also have to update the index and mightn't know to). Unless I'm missing a property of rglob that I can use here?

Reply:
I think glob patterns can do exclusion, although I've never done it myself. If you have a look and find that it's nasty then by all means keep this bit (although I still think `rglob()` will be a good replacement for `getFilepaths()`).

Reply:
It was either not possible or not pretty so I've left these in a list, though I've dragged them out so that the functions are more modular